finetune/target_models/classifier.py

```diff
@@ -51,6 +72,169 @@ def __init__(self, **kwargs):
+
+    @classmethod
+    def get_eval_fn(cls):
+        return lambda labels, targets: np.mean(
+            np.asarray(labels) == np.asarray(targets)
+        )
+
+    def _target_model(
+        self, *, config, featurizer_state, targets, n_outputs, train=False, reuse=None, **kwargs
+    ):
+        if "explain_out" in featurizer_state:
+            shape = tf.shape(featurizer_state["explain_out"])  # batch, seq, hidden
+            flat_explain = tf.reshape(
+                featurizer_state["explain_out"], [shape[0] * shape[1], shape[2]]
+            )
+            hidden = tf.concat((featurizer_state["features"], flat_explain), 0)
+        else:
+            hidden = featurizer_state["features"]
+
+        clf_out = classifier(
+            hidden=hidden,
+            targets=targets,
+            n_targets=n_outputs,
+            config=config,
+            train=train,
+            reuse=reuse,
+            **kwargs
+        )
+        if "explain_out" in featurizer_state:
+            logits = clf_out["logits"]
+            clf_out["logits"] = logits[: shape[0]]
+            clf_out["explanation"] = tf.nn.softmax(
+                tf.reshape(
+                    logits[shape[0] :],
+                    tf.concat((shape[:2], [tf.shape(logits)[-1]]), 0),
+                ),
+                -1,
+            )
+        return clf_out
+
+    def explain(self, Xs):
+        explanation = self._inference(
+            Xs, predict_keys=[PredictMode.EXPLAIN, PredictMode.NORMAL]
+        )
+        classes = self.input_pipeline.label_encoder.target_labels
+        out = []
+        bases = []
+        preds = []
+        for values in explanation:
+            explain_sample = values[PredictMode.EXPLAIN]
+            preds.append(values[PredictMode.NORMAL])
+            out.append(explain_sample[1:])
+            bases.append(explain_sample[0])
+        processed = finetune_to_indico_explain(
+            Xs, out, self.input_pipeline.text_encoder, attention=False
+        )
+
+        for base, sample, cls in zip(bases, processed, preds):
+            weights = sample["explanation"]
+            weights = np.array([base] + weights[:-1]) - weights
+            n_classes = weights.shape[-1]
+            norm = (
+                np.max([np.abs(np.max(weights, 0)), abs(np.min(weights, 0))], 0)
+                * n_classes
+            )
+            explanation = weights / norm + 1 / n_classes
+
+            sample["explanation"] = {
+                c: explanation[:, i] for i, c in enumerate(classes)
+            }
+            sample["prediction"] = self.input_pipeline.label_encoder.inverse_transform(
+                [cls]
+            )[0]
+
+        return processed
+
+    def _predict_op(self, logits, **kwargs):
+        return tf.contrib.seq2seq.hardmax(logits)
+
+    def _predict_proba_op(self, logits, **kwargs):
+        return tf.nn.softmax(logits, -1)
+
+class NoisyClassifier(BaseModel):
```

Comment:
Inherit from classifier and just override required methods

Reply:
You got it!